### DISCUSSION TITLE
Is it OK to read a provider using GlobalKey?

### DISCUSSION THREAD

**Discussion by XuanTung95** [rrousselGit/riverpod#1500] [General]
Idea: 
Using `GlobalKey` to obtain a BuildContext. 
Using `ProviderScope.containerOf(context, listen: false).read(provider)` instead of ref.read

Use case: Doing something globally without ref.

Code:

```
import 'package:flutter/material.dart';
import 'package:flutter_riverpod/flutter_riverpod.dart';

final appKey = GlobalKey();
final provider = StateProvider((ref) => 0);

void main() {
  runApp(
    ProviderScope(
      child: MaterialApp(
        key: appKey,
        home: HomePage(),
      ),
    ),
  );
}

class HomePage extends StatelessWidget {
  const HomePage({Key? key}) : super(key: key);

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      body: Center(
        child: ElevatedButton(
          onPressed: () {
            if (appKey.currentContext != null) {
              final state = ProviderScope.containerOf(appKey.currentContext!, listen: false).read(provider.notifier);
              state.state++;
            }
          },
          child: Text('Do things'),
        ),
      ),
    );
  }
}
```

**Comment by tbm98** [comment]
[riverpod-initialize-listener-app-startup](https://codewithandrea.com/articles/riverpod-initialize-listener-app-startup/#conclusion) may help you.

**Reply by XuanTung95** [reply]
You are right. Creating ProviderContainer() manually is another approach for this. I wonder which approach is recommended and safe to use.

**Reply by tbm98** [reply]
I think your way is not a problem. The consumer widget also uses ProviderScope.containerOf function by itself.
